<a href="https://colab.research.google.com/github/MUMADE-TADM/s1_python-manunu93/blob/main/2_Copia_de_TADM2022_S03b_Consultas_SQL_(Clase)_Manuel_V%C3%ADctor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Consultas SQL


Antes de empezar cargamos nuestro archivo data_berka en nuestro entorno: [enlace](https://sorry.vse.cz/~berka/challenge/pkdd1999/berka.htm)

In [ ]:
!gdown --id 1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU #bajamos la base de datos

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1Osmr7qVN0NMuqYxvLesNn7H4nqPCVbsU
To: /content/data_berka.db
100% 70.2M/70.2M [00:01<00:00, 62.6MB/s]


En primer lugar cargamos la extensión que nos va permitir manejar comandos sql directamente en colab


cargar sentencia para hacer llamadas con el lenguaje sql

In [ ]:
%load_ext sql

Y conectamos con la base de datos

In [ ]:
%sql sqlite:///data_berka.db 

'Connected: @data_berka.db'

Comprobamos esquema

In [ ]:
%%sql
SELECT name,type
FROM sqlite_master
WHERE type = 'table'

 * sqlite:///data_berka.db
Done.


[('DISTRICT', 'table'),
 ('ACCOUNT', 'table'),
 ('TRANS', 'table'),
 ('CLIENT', 'table'),
 ('DISPOSITION', 'table'),
 ('CREDIT_CARD', 'table'),
 ('LOAN', 'table'),
 ('PERMANENT_ORDER', 'table')]

## Consultas sobre Préstamos

1. Obtener todos los prestamos que tiene el banco

(extra limitar a 20 registros)

In [ ]:
# un % para una linea dos %% para toda la celda sea sql

%%sql select * 
from LOAN
LIMIT 20

 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B')]

2. Obtener los campos de fecha (date), amount (cantidad), duración (duration), pagos (payments), estado (estado)

(extra limitar a 20 registros)

In [ ]:
%%sql 
select date, amount, duration, payments, status
from LOAN
order by amount desc
LIMIT 20


 * sqlite:///data_berka.db
Done.


[(971019, 590820, 60, 9847, 'C'),
 (980123, 566640, 60, 9444, 'C'),
 (971112, 541200, 60, 9020, 'D'),
 (950217, 538500, 60, 8975, 'C'),
 (980120, 504000, 60, 8400, 'C'),
 (980522, 495180, 60, 8253, 'C'),
 (941219, 482940, 60, 8049, 'D'),
 (970212, 475680, 48, 9910, 'C'),
 (970907, 473280, 60, 7888, 'D'),
 (971225, 468060, 60, 7801, 'C'),
 (970613, 466608, 48, 9721, 'C'),
 (971227, 465504, 48, 9698, 'D'),
 (950709, 465072, 48, 9689, 'C'),
 (931201, 464520, 60, 7742, 'B'),
 (971105, 460980, 60, 7683, 'D'),
 (960723, 444864, 48, 9268, 'C'),
 (980829, 437460, 60, 7291, 'C'),
 (980213, 428784, 48, 8933, 'C'),
 (961118, 421008, 48, 8771, 'C'),
 (951029, 419880, 60, 6998, 'C')]

3. Obtener los distintos tipos de estado que se presentan en un préstamo.


para el order by se puede poner argumentos extra para ponerlo ascendente o descendente

In [ ]:
## el distinct se usa para obtener los valores distintos que estan en una columna de una tabla

%%sql 
select distinct status
from LOAN
order by status 


 * sqlite:///data_berka.db
Done.


[('A',), ('B',), ('C',), ('D',)]

In [ ]:
%%sql 
select distinct status
from LOAN


 * sqlite:///data_berka.db
Done.


[('B',), ('A',), ('C',), ('D',)]

In [ ]:
# con esta query se consumen más recursos
%%sql 
select status
from LOAN
group by status

 * sqlite:///data_berka.db
Done.


[('A',), ('B',), ('C',), ('D',)]

4. Obtener solo los prestamos ya terminados.

In [ ]:
%%sql
select loan_id as préstamo, status as estado
from LOAN
where estado in ("A", "B") -- where siempre antes de un group by, having siempre despues
limit 10 

 * sqlite:///data_berka.db
Done.


[(5314, 'B'),
 (5316, 'A'),
 (6863, 'A'),
 (5325, 'A'),
 (7240, 'A'),
 (6687, 'A'),
 (7284, 'A'),
 (6111, 'B'),
 (7235, 'A'),
 (5997, 'A')]

Tener muy presente siempre la estructura y la leyenda de la base de datos ya que algunas query pueden y van a ser combinaciones de varias tablas de la base de datos

In [ ]:
%%sql
select * 
from LOAN
where status in ("A", "B")


 * sqlite:///data_berka.db
Done.


[(5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (5997, 4894, 931104, 117024, 24, 4876, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (6077, 5270, 931122, 79608, 24, 3317, 'A'),
 (6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (6356, 6701, 931208, 95400, 36, 2650, 'A'),
 (5523, 2705, 931208, 93888, 36, 2608, 'A'),
 (6456, 7123, 931209, 47016, 12, 3918, 'A'),
 (7104, 10320, 931213, 259740, 60, 4329, 'A'),
 (6820, 9034, 931216, 38148, 12, 3179, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A'),
 (7097, 10266, 931223, 75624, 24, 3151, 'B'),
 (4959, 2, 940105, 80952, 24, 3373, 'A'),
 (6499, 7401, 940105, 80952, 24, 3373, 'A'),
 (5479, 2486, 940110, 24516, 12, 2043, 'A'),
 (6721, 8475, 940121, 21072, 24, 878, 'A'),
 (7259, 11111, 940131, 108144, 36, 3004, 'A'),
 (5285, 1603, 940206, 78936, 12, 6578, 'A'),
 (5189, 1166, 940207, 149040, 48, 3105, 'A'),
 (6650, 8073, 940217, 49320, 12, 4110, 'B'),
 (7226, 10940, 940223, 197748, 36, 5493, 'A'),
 (7262, 11135, 940301, 182628, 36, 5073, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (6577, 7753, 940311, 51696, 24, 2154, 'A'),
 (6509, 7445, 940315, 55368, 24, 2307, 'A'),
 (6520, 7485, 940316, 91200, 24, 3800, 'A'),
 (6699, 8330, 940330, 7656, 24, 319, 'A'),
 (6336, 6555, 940331, 212400, 36, 5900, 'A'),
 (6202, 5891, 940418, 65184, 12, 5432, 'A'),
 (7066, 10131, 940502, 215388, 36, 5983, 'B'),
 (6072, 5256, 940510, 193968, 36, 5388, 'A'),
 (5130, 813, 940511, 24312, 12, 2026, 'A'),
 (6678, 8173, 940520, 80340, 12, 6695, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6580, 7769, 940522, 196080, 48, 4085, 'A'),
 (6316, 6473, 940526, 76908, 12, 6409, 'B'),
 (6512, 7454, 940526, 63312, 48, 1319, 'A'),
 (6407, 6930, 940526, 25944, 24, 1081, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (6647, 8051, 940601, 208320, 48, 4340, 'A'),
 (7137, 10439, 940603, 151560, 24, 6315, 'A'),
 (6903, 9337, 940604, 52128, 24, 2172, 'A'),
 (5918, 4576, 940604, 300204, 36, 8339, 'A'),
 (6642, 8027, 940605, 77640, 12, 6470, 'A'),
 (6210, 5927, 940606, 87216, 48, 1817, 'B'),
 (6279, 6273, 940607, 87360, 24, 3640, 'A'),
 (7249, 11065, 940612, 69360, 48, 1445, 'A'),
 (6112, 5442, 940617, 174744, 24, 7281, 'B'),
 (6596, 7844, 940625, 91584, 24, 3816, 'A'),
 (5657, 3354, 940705, 4980, 12, 415, 'A'),
 (5533, 2762, 940705, 111384, 24, 4641, 'A'),
 (6232, 6051, 940708, 276660, 36, 7685, 'B'),
 (6888, 9265, 940711, 39576, 12, 3298, 'A'),
 (6066, 5228, 940711, 173808, 24, 7242, 'A'),
 (6004, 4937, 940723, 143904, 24, 5996, 'A'),
 (5126, 790, 940724, 208128, 48, 4336, 'B'),
 (7220, 10911, 940728, 131292, 36, 3647, 'A'),
 (6668, 8129, 940730, 74736, 36, 2076, 'A'),
 (5395, 2176, 940730, 299088, 36, 8308, 'B'),
 (6064, 5215, 940804, 14628, 12, 1219, 'A'),
 (6919, 9403, 940805, 120888, 36, 3358, 'A'),
 (6841, 9104, 940807, 172080, 48, 3585, 'A'),
 (6038, 5120, 940810, 323472, 48, 6739, 'A'),
 (6729, 8519, 940811, 74688, 12, 6224, 'A'),
 (7198, 10788, 940823, 45768, 12, 3814, 'A'),
 (6686, 8260, 940825, 224604, 36, 6239, 'A'),
 (5836, 4259, 940827, 79632, 24, 3318, 'A'),
 (7061, 10105, 940827, 121896, 24, 5079, 'A'),
 (5595, 3050, 940902, 82896, 12, 6908, 'A'),
 (5723, 3674, 940910, 23628, 12, 1969, 'A'),
 (6120, 5481, 940911, 209184, 48, 4358, 'A'),
 (6895, 9307, 940919, 41904, 12, 3492, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (5754, 3834, 940928, 23052, 12, 1921, 'A'),
 (6594, 7824, 940928, 189696, 24, 7904, 'B'),
 (6738, 8566, 940929, 230220, 36, 6395, 'B'),
 (6624, 7965, 940930, 71460, 12, 5955, 'A'),
 (6216, 5952, 940930, 141240, 24, 5885, 'A'),
 (6229, 6040, 941002, 11736, 24, 489, 'A'),
 (5887, 4448, 941003, 192744, 36, 5354, 'A'),

5. Ordenar lo anterior por duración y cantidad

extra: limitar a 20 elementos

In [ ]:
%%sql
select * 
from LOAN
where status in ("A", "B")
order by duration, amount desc -- el comando por defecto es asc, pero con en atributo desc se invierte
limit 20

 * sqlite:///data_berka.db
Done.


[(6934, 9494, 970209, 116832, 12, 9736, 'A'),
 (5002, 173, 940531, 104808, 12, 8734, 'A'),
 (4986, 97, 970810, 102876, 12, 8573, 'A'),
 (5375, 2106, 961003, 101520, 12, 8460, 'A'),
 (7253, 11079, 970426, 98304, 12, 8192, 'A'),
 (6998, 9833, 940521, 97632, 12, 8136, 'A'),
 (6612, 7907, 970831, 97392, 12, 8116, 'A'),
 (5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (6995, 9814, 960926, 96168, 12, 8014, 'A'),
 (5464, 2419, 950130, 95808, 12, 7984, 'A'),
 (5643, 3273, 970921, 94512, 12, 7876, 'B'),
 (5377, 2116, 970804, 92400, 12, 7700, 'A'),
 (5519, 2694, 940308, 91632, 12, 7636, 'A'),
 (7046, 10049, 970328, 91632, 12, 7636, 'A'),
 (5868, 4361, 970329, 90156, 12, 7513, 'A'),
 (7277, 11231, 971024, 89280, 12, 7440, 'A'),
 (4996, 132, 961106, 88440, 12, 7370, 'A'),
 (5882, 4427, 940926, 87816, 12, 7318, 'A'),
 (7119, 10361, 970602, 86880, 12, 7240, 'A'),
 (6876, 9236, 931221, 86616, 12, 7218, 'A')]

In [ ]:
#importar pandas python
import pandas as pd

In [ ]:
#cargar resultados de un DF de padas
df = %sql select *  from LOAN where status in ("A", "B") order by duration, amount desc

#conversión a csv
df.DataFrame().to_csv("first_sql.csv")

 * sqlite:///data_berka.db
Done.


In [ ]:
df.DataFrame()

loan_id  account_id    date  amount  duration  payments status
0       6934        9494  970209  116832        12      9736      A
1       5002         173  940531  104808        12      8734      A
2       4986          97  970810  102876        12      8573      A
3       5375        2106  961003  101520        12      8460      A
4       7253       11079  970426   98304        12      8192      A
..       ...         ...     ...     ...       ...       ...    ...
229     6223        5991  941103   61056        48      1272      A
230     6228        6034  931201  464520        60      7742      B
231     7240       11013  930906  274740        60      4579      A
232     7104       10320  931213  259740        60      4329      A
233     6863        9188  930728  127080        60      2118      A

[234 rows x 7 columns]

Ex. 1

>Plantear otro filtro para los préstamos (por cantidad, duración, status, etc.)


In [ ]:
%%sql
select * 
from LOAN
where status in ("A", "B")
order by  amount, duration, status desc
limit 20

 * sqlite:///data_berka.db
Done.


[(5657, 3354, 940705, 4980, 12, 415, 'A'),
 (6699, 8330, 940330, 7656, 24, 319, 'A'),
 (6019, 5033, 960725, 11400, 12, 950, 'A'),
 (6229, 6040, 941002, 11736, 24, 489, 'A'),
 (5103, 666, 971128, 12540, 12, 1045, 'A'),
 (5718, 3637, 970901, 14028, 12, 1169, 'A'),
 (6064, 5215, 940804, 14628, 12, 1219, 'A'),
 (5897, 4476, 960925, 15192, 24, 633, 'A'),
 (5599, 3064, 950321, 15420, 12, 1285, 'A'),
 (6377, 6792, 970317, 17184, 12, 1432, 'A'),
 (6297, 6363, 941025, 17508, 12, 1459, 'A'),
 (6942, 9513, 970716, 18324, 12, 1527, 'A'),
 (5740, 3752, 950402, 18564, 12, 1547, 'A'),
 (5651, 3329, 971208, 19248, 12, 1604, 'A'),
 (5922, 4580, 950805, 20220, 12, 1685, 'A'),
 (6721, 8475, 940121, 21072, 24, 878, 'A'),
 (5980, 4825, 950818, 21396, 12, 1783, 'A'),
 (6600, 7861, 951008, 21852, 36, 607, 'A'),
 (7121, 10364, 931110, 21924, 36, 609, 'A'),
 (5754, 3834, 940928, 23052, 12, 1921, 'A')]

 Ex. 2 

> Seleccionar fecha, cantidad, duración y pagos de aquellos préstamos que han sido terminados con una duración mayor 12 meses ordenados de mayor a menor cantidad de dinero



In [ ]:
%%sql
select date, amount, duration, status, payments
from LOAN
where status in ("A", "B") 
and duration > 12
order by amount desc
limit 10

 * sqlite:///data_berka.db
Done.


[(931201, 464520, 60, 'B', 7742),
 (940810, 323472, 48, 'A', 6739),
 (940604, 300204, 36, 'A', 8339),
 (940730, 299088, 36, 'B', 8308),
 (950103, 288468, 36, 'A', 8013),
 (940708, 276660, 36, 'B', 7685),
 (930906, 274740, 60, 'A', 4579),
 (950108, 270648, 36, 'B', 7518),
 (950623, 265320, 36, 'A', 7370),
 (951028, 260028, 36, 'A', 7223)]

Cuantos registros cumplen esa codición

In [ ]:
%%sql
select count(*)
from LOAN
where status in ("A", "B") and duration > 12



 * sqlite:///data_berka.db
Done.


[(131,)]

¿Cuantos préstamos no terminados con duración  entre 24 y 48

In [ ]:
%%sql
select count(*)
from LOAN
where status in ("C", "D") and duration >= 24 and duration <= 48 

 * sqlite:///data_berka.db
Done.


[(279,)]

In [ ]:
%%sql
select count(*)
from LOAN
where status in ("C", "D") and duration between 24 and 48 -- se puede poner de varias formas una restricción

 * sqlite:///data_berka.db
Done.


[(279,)]

## Group By

7. ¿Cuántos préstamos hay por cada estado y cual es la media de la cantidad de dinero prestado?

In [ ]:
%%sql
select status, count(*) total, round(avg(amount), 2) as "Media prestado"
from LOAN
group by status -- comando para agrupar terminos
order by total desc

 * sqlite:///data_berka.db
Done.


[('C', 403, 171410.35),
 ('A', 203, 91641.46),
 ('D', 45, 249284.53),
 ('B', 31, 140720.9)]

In [ ]:
%%sql
select status, count(*) total, round(avg(amount), 2) as "Media prestado"
from LOAN
where duration >24
group by status
order by total desc


 * sqlite:///data_berka.db
Done.


[('C', 319, 195017.83),
 ('A', 46, 163045.3),
 ('D', 38, 277804.42),
 ('B', 10, 227607.6)]

In [ ]:
%%sql
select status, count(*) total, round(avg(amount), 2) as "Media prestado"
from LOAN
where duration >24
group by status
having total >25
order by total desc

 * sqlite:///data_berka.db
Done.


[('C', 319, 195017.83), ('A', 46, 163045.3), ('D', 38, 277804.42)]

El where se utiliza para filtrar antes de agrupaciones y el having despues de agrupaciones

8. Por cada Región, quiero contar el número de ciudades y obtener el total del número de habitantes de la región

extra: no cosiderar la region de Praga


district
a3
a9
a4: total de habitantes

In [ ]:
%%sql
select a3 as region, count(a2) as "n_districtos", sum(a9) as "n_cities", sum(a4) as "sum_habitantes"
from district
group by a3
having region != "Prague" -- operador para diferente es el igual y la exclamación

 * sqlite:///data_berka.db
Done.


[('central Bohemia', 12, 66, 1105234),
 ('east Bohemia', 11, 86, 1234781),
 ('north Bohemia', 10, 70, 1178977),
 ('north Moravia', 11, 60, 1970302),
 ('south Bohemia', 8, 48, 700595),
 ('south Moravia', 14, 82, 2054989),
 ('west Bohemia', 10, 69, 859306)]

9. Contar ordenes permanentes (domiciliaciones) por banco y tipo de transacción

In [ ]:
%%sql
select bank_to, k_symbol, count(order_id)
from permanent_order
group by bank_to, k_symbol
order by bank_to, k_symbol
limit 20

 * sqlite:///data_berka.db
Done.


[('AB', ' ', 113),
 ('AB', 'LEASING', 24),
 ('AB', 'POJISTNE', 46),
 ('AB', 'SIPO', 278),
 ('AB', 'UVER', 58),
 ('CD', ' ', 94),
 ('CD', 'LEASING', 23),
 ('CD', 'POJISTNE', 28),
 ('CD', 'SIPO', 263),
 ('CD', 'UVER', 50),
 ('EF', ' ', 96),
 ('EF', 'LEASING', 22),
 ('EF', 'POJISTNE', 46),
 ('EF', 'SIPO', 267),
 ('EF', 'UVER', 52),
 ('GH', ' ', 102),
 ('GH', 'LEASING', 22),
 ('GH', 'POJISTNE', 43),
 ('GH', 'SIPO', 260),
 ('GH', 'UVER', 60)]

# JOIN

10. ¿Cuántas cuentas de frecuencia mensual tienen un un préstamo activo?



hacerlo por inner join o hacerlo por where es lo mismo

si se mete un + al lado del igual seria como un left o un right join 

In [ ]:
#Query par saber las cuentas con frecuencia mensual y diversos datos
%%sql
select loan_id, a.date loan_date, amount, duration, payments, frequency, b.date acc_date
from loan a, account b 
where a.account_id = b.account_id
and status in ("A", "B")
and frequency is "monthly" -- frecuencia 
order by amount desc, a.date desc
limit 12

 * sqlite:///data_berka.db
Done.


[(6038, 940810, 323472, 48, 6739, 'monthly', 930621),
 (5918, 940604, 300204, 36, 8339, 'monthly', 930605),
 (5395, 940730, 299088, 36, 8308, 'monthly', 930606),
 (6526, 950103, 288468, 36, 8013, 'monthly', 940829),
 (6232, 940708, 276660, 36, 7685, 'monthly', 930828),
 (6725, 950623, 265320, 36, 7370, 'monthly', 930918),
 (7176, 951028, 260028, 36, 7223, 'monthly', 931208),
 (7104, 931213, 259740, 60, 4329, 'monthly', 930711),
 (5310, 950428, 248148, 36, 6893, 'monthly', 931211),
 (6738, 940929, 230220, 36, 6395, 'monthly', 931008),
 (6808, 941012, 227880, 36, 6330, 'monthly', 930925),
 (6686, 940825, 224604, 36, 6239, 'monthly', 930324)]

In [ ]:
#Query par saber los préstamos con frecuencia mensual
%%sql
select count(*)
from loan a, account b 
where a.account_id = b.account_id
and status in ("A", "B")
and frequency is "monthly" -- frecuencia 
order by amount desc, a.date desc
limit 12

 * sqlite:///data_berka.db
Done.


[(190,)]

In [ ]:
#número de préstamos de solo usuarios
%%sql
select count(*)
from loan l, account a, DISTRICT d, disposition dp, client c
where l.account_id = a.account_id
and a.district_id = d.A1
and dp.account_id = a.account_id
and dp.client_id = c.client_id
and dp.type ="OWNER" -- tipo de disposición
and frequency is "monthly"
and status in ("A", "B")
order by amount desc, a.date desc
limit 12

 * sqlite:///data_berka.db
Done.


[(190,)]

In [ ]:
#tipo de disposición sin filtrar por la frecuencia
%%sql
select loan_id, l.date loan_date, amount, duration, payments, frequency, a.date acc_date, l.date loan_date, amount, duration
from loan l, account a, DISTRICT d, disposition dp, client c
where l.account_id = a.account_id
and a.district_id = d.A1
and dp.account_id = a.account_id
and dp.client_id = c.client_id
and dp.type ="OWNER"
and status in ("A", "B")
order by amount desc, a.date desc
limit 12

 * sqlite:///data_berka.db
Done.


[(6228, 931201, 464520, 60, 7742, 'weekly', 930319, 931201, 464520, 60),
 (6038, 940810, 323472, 48, 6739, 'monthly', 930621, 940810, 323472, 48),
 (5918, 940604, 300204, 36, 8339, 'monthly', 930605, 940604, 300204, 36),
 (5395, 940730, 299088, 36, 8308, 'monthly', 930606, 940730, 299088, 36),
 (6526, 950103, 288468, 36, 8013, 'monthly', 940829, 950103, 288468, 36),
 (6232, 940708, 276660, 36, 7685, 'monthly', 930828, 940708, 276660, 36),
 (7240, 930906, 274740, 60, 4579, 'weekly', 930214, 930906, 274740, 60),
 (6244, 950108, 270648, 36, 7518, 'weekly', 940307, 950108, 270648, 36),
 (6725, 950623, 265320, 36, 7370, 'monthly', 930918, 950623, 265320, 36),
 (7176, 951028, 260028, 36, 7223, 'monthly', 931208, 951028, 260028, 36),
 (7104, 931213, 259740, 60, 4329, 'monthly', 930711, 931213, 259740, 60),
 (5310, 950428, 248148, 36, 6893, 'monthly', 931211, 950428, 248148, 36)]

In [ ]:
## query que se ha añadido un substr para obtener los años
%%sql
select loan_id, l.date loan_date, amount, substr(c.birth_number, 0,3) birthyear, duration, payments, frequency, a.date acc_date, l.date loan_date, amount, duration
from loan l, account a, DISTRICT d, disposition dp, client c
where l.account_id = a.account_id
and a.district_id = d.A1
and dp.account_id = a.account_id
and dp.client_id = c.client_id
and dp.type ="OWNER"
and status in ("A", "B")
order by amount desc, a.date desc
limit 12

 * sqlite:///data_berka.db
Done.


[(6228, 931201, 464520, '40', 60, 7742, 'weekly', 930319, 931201, 464520, 60),
 (6038, 940810, 323472, '40', 48, 6739, 'monthly', 930621, 940810, 323472, 48),
 (5918, 940604, 300204, '64', 36, 8339, 'monthly', 930605, 940604, 300204, 36),
 (5395, 940730, 299088, '66', 36, 8308, 'monthly', 930606, 940730, 299088, 36),
 (6526, 950103, 288468, '58', 36, 8013, 'monthly', 940829, 950103, 288468, 36),
 (6232, 940708, 276660, '49', 36, 7685, 'monthly', 930828, 940708, 276660, 36),
 (7240, 930906, 274740, '78', 60, 4579, 'weekly', 930214, 930906, 274740, 60),
 (6244, 950108, 270648, '37', 36, 7518, 'weekly', 940307, 950108, 270648, 36),
 (6725, 950623, 265320, '48', 36, 7370, 'monthly', 930918, 950623, 265320, 36),
 (7176, 951028, 260028, '76', 36, 7223, 'monthly', 931208, 951028, 260028, 36),
 (7104, 931213, 259740, '79', 60, 4329, 'monthly', 930711, 931213, 259740, 60),
 (5310, 950428, 248148, '40', 36, 6893, 'monthly', 931211, 950428, 248148, 36)]

11. Domiciliaciones en cuentas de actualización semanal de alquiler en cuentas de Praga

Region(a3)


In [ ]:
%%sql
select ac.account_id, po.bank_to,  po.k_symbol, ac.frequency, di.a3
from permanent_order po , account ac, district di
where po.account_id = ac.account_id
and ac.district_id = di.a1
and di.a3 = "Prague"
and ac.frequency = "weekly"
and po.k_symbol = "SIPO"
limit 5

 * sqlite:///data_berka.db
Done.


[(72, 'EF', 'SIPO', 'weekly', 'Prague'),
 (179, 'UV', 'SIPO', 'weekly', 'Prague'),
 (186, 'GH', 'SIPO', 'weekly', 'Prague'),
 (510, 'YZ', 'SIPO', 'weekly', 'Prague'),
 (619, 'YZ', 'SIPO', 'weekly', 'Prague')]


12. Domiciliaciones (*órdenes*) semanales para alquiler (*leasing*) en cuentas de Praga

 - Ordenes permanentes de alquiler (*leasing*) y semanales (*weekly*)
 - Relación con cuentas que tienen región Praga

In [ ]:
%%sql
select ac.account_id, po.bank_to,  po.k_symbol, ac.frequency, di.a3
from permanent_order po , account ac, district di
where po.account_id = ac.account_id
and ac.district_id = di.a1
and di.a3 = "Prague"
and ac.frequency = "weekly"
and po.k_symbol = "LEASING"
limit 5

 * sqlite:///data_berka.db
Done.


[(72, 'OP', 'LEASING', 'weekly', 'Prague'),
 (573, 'CD', 'LEASING', 'weekly', 'Prague'),
 (1056, 'IJ', 'LEASING', 'weekly', 'Prague'),
 (3141, 'OP', 'LEASING', 'weekly', 'Prague'),
 (3959, 'AB', 'LEASING', 'weekly', 'Prague')]

13. Transacciones relacionadas con pensiones (DUCHOD) en Plzen desde el banco MN

(limitar a 10)

In [ ]:
#nº de transacción, el banco y el tipo de transacción y la región, como la región tiene varias zonas
#se usa el like para indicar que incluyan ese nombre al principio
%%sql

select tr.trans_id as transacción, tr.bank as banco ,tr.k_symbol as tipo,  di.A2 as region
from TRANS tr, ACCOUNT ac, DISTRICT di
where tr.account_id = ac.account_id
and ac.district_id = di.A1
and tr.k_symbol is "DUCHOD"
and tr.bank = "MN" --se puede hacer por is o por iguales
and A2 like 'Plzen%' 
limit 10

 * sqlite:///data_berka.db
Done.


[(47969, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47970, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47971, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47972, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47973, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47974, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47975, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47976, 'MN', 'DUCHOD', 'Plzen - sever'),
 (47977, 'MN', 'DUCHOD', 'Plzen - sever'),
 (1098524, 'MN', 'DUCHOD', 'Plzen - sever')]

# Subconsultas

14. Prestamos cuya cantidad es superior a la media de los clientes de Praga

- Cantidad media prestada a los cuentas de Praga 

In [ ]:
%%sql
select avg(l.amount)
from loan l, account a, district d
where l.account_id = a.account_id and a.district_id = d.a1
and d.a3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(153957.2857142857,)]

- Consultar los préstamos superiores a esa cantidad

In [ ]:
%%sql
select count(*)
from loan
where amount > (select avg(l.amount)
                    from loan l, account a, district d
                    where l.account_id = a.account_id and a.district_id = d.a1
                    and d.a3 = "Prague")




 * sqlite:///data_berka.db
Done.


[(277,)]

# Repaso de Join/Group By

## Objetivo 2.1

15 . Clientes de Praga con préstamos según su estado

- ¿Cuántos clientes hay en Praga?
- ¿Cuántas cuentas tienen asociados clientes en Praga?
- ¿cuántas de esas cuentas tienen asociados un préstamo?
- Necesito saber cuántos hay por cada estado (*status*)


In [ ]:
#a. ¿Cuántos clientes hay en Praga?

%%sql

select count(*)
from  DISTRICT di, CLIENT cl
where di.A1 = cl.district_id
and di.A3 = "Prague"


 * sqlite:///data_berka.db
Done.


[(663,)]

In [ ]:
#b. ¿Cuántas cuentas tienen asociados clientes en Praga? -> entiendo que cuantas cuentas hay en total en praga

%%sql

select count(*)
from  DISTRICT di, ACCOUNT ac
where di.A1 = ac.district_id
and di.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(554,)]

In [ ]:
#c. ¿cuántas de esas cuentas tienen asociados un préstamo?

%%sql

select count(*)
from  DISTRICT di, ACCOUNT ac, loan lo
where di.A1 = ac.district_id
and ac.account_id = lo.account_id
and di.A3 = "Prague"

 * sqlite:///data_berka.db
Done.


[(84,)]

In [ ]:
#d. Necesito saber cuántos hay por cada estado (status)

%%sql

select lo.status, count(status)
from  DISTRICT di, ACCOUNT ac, loan lo
where di.A1 = ac.district_id
and ac.account_id = lo.account_id
and di.A3 = "Prague"
group by lo.status

 * sqlite:///data_berka.db
Done.


[('A', 34), ('B', 3), ('C', 43), ('D', 4)]

## Ejercicio

> cantidad de dinero medio en transacciones de clientes propietarios de cuentas en Kolin (owned accounts)

In [ ]:
%%sql
SELECT c.client_id, avg (amount) as av
from client c, district d, disposition ds, account a, trans t
WHERE c.district_id = d.A1 and c.client_id = ds.client_id
    and ds.account_id = a.account_id and a.account_id = t.account_id
and d.A2 = "Kolin" and ds.type = "OWNER" 
GROUP BY c.client_id
ORDER BY av DESC
LIMIT 10



 * sqlite:///data_berka.db
Done.


[(5249, 15611.831578947369),
 (3378, 14803.997402597402),
 (1388, 14464.587096774194),
 (6803, 13825.125153374232),
 (8091, 13227.722338204592),
 (3298, 13074.464857881138),
 (1806, 13037.759375),
 (4331, 12644.448051948051),
 (4974, 12023.291743119267),
 (3080, 11784.50460251046)]

> sanction interest if negative balance Cesky Krumlov


> stands for loan payment in Praha from the bank EF

# Extra: Left Join

1. ¿Cuántas cuentas hay? 4500
2. ¿Cuántas cuentas hay con préstamos? (inner join) 682
3. Cuentas con los datos de los préstamos si los hay 234
4. los ordenamos por cantidad


In [ ]:
#a. ¿Cuántas cuentas hay?
%%sql 
select count(*)
from account




 * sqlite:///data_berka.db
Done.


[(4500,)]

In [ ]:
#b. ¿Cuántas cuentas hay con préstamos? (inner join)
%%sql

select count(*)
from account ac, loan lo
where ac.account_id =lo.account_id



 * sqlite:///data_berka.db
Done.


[(682,)]

In [ ]:
#c. Cuentas con los datos de los préstamos si los hay

%%sql

select *
from account ac, loan lo
where ac.account_id =lo.account_id
and status in ("A", "B")
limit 10

 * sqlite:///data_berka.db
Done.


[(1787, 30, 'weekly', 930322, 5314, 1787, 930705, 96396, 12, 8033, 'B'),
 (1801, 46, 'monthly', 930213, 5316, 1801, 930711, 165960, 36, 4610, 'A'),
 (9188, 45, 'monthly', 930208, 6863, 9188, 930728, 127080, 60, 2118, 'A'),
 (1843, 12, 'monthly', 930130, 5325, 1843, 930803, 105804, 36, 2939, 'A'),
 (11013, 1, 'weekly', 930214, 7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (8261, 13, 'monthly', 930512, 6687, 8261, 930913, 87840, 24, 3660, 'A'),
 (11265, 15, 'monthly', 930114, 7284, 11265, 930915, 52788, 12, 4399, 'A'),
 (5428, 24, 'monthly', 930414, 6111, 5428, 930924, 174744, 24, 7281, 'B'),
 (10973, 18, 'weekly', 930420, 7235, 10973, 931013, 154416, 48, 3217, 'A'),
 (4894, 14, 'monthly', 930620, 5997, 4894, 931104, 117024, 24, 4876, 'A')]

In [ ]:
%%sql

select count(*)
from account ac, loan lo
where ac.account_id =lo.account_id
and status in ("A", "B")
limit 10

 * sqlite:///data_berka.db
Done.


[(234,)]

In [ ]:
#d. los ordenamos por cantidad
%%sql


select *
from account ac, loan lo
where ac.account_id = lo.account_id
and status in ("A", "B")
order by amount desc
limit 10



 * sqlite:///data_berka.db
Done.


[(6034, 6, 'weekly', 930319, 6228, 6034, 931201, 464520, 60, 7742, 'B'),
 (5120, 1, 'monthly', 930621, 6038, 5120, 940810, 323472, 48, 6739, 'A'),
 (4576, 27, 'monthly', 930605, 5918, 4576, 940604, 300204, 36, 8339, 'A'),
 (2176, 67, 'monthly', 930606, 5395, 2176, 940730, 299088, 36, 8308, 'B'),
 (7523, 63, 'monthly', 940829, 6526, 7523, 950103, 288468, 36, 8013, 'A'),
 (6051, 56, 'monthly', 930828, 6232, 6051, 940708, 276660, 36, 7685, 'B'),
 (11013, 1, 'weekly', 930214, 7240, 11013, 930906, 274740, 60, 4579, 'A'),
 (6118, 1, 'weekly', 940307, 6244, 6118, 950108, 270648, 36, 7518, 'B'),
 (8489, 19, 'monthly', 930918, 6725, 8489, 950623, 265320, 36, 7370, 'A'),
 (10663, 11, 'monthly', 931208, 7176, 10663, 951028, 260028, 36, 7223, 'A')]

# DataSets

---

Dataset sobre las transacciones con Vyskov, extrayendo los años de las diferentes fechas. 

In [ ]:
%%sql dataset << SELECT substr(a.date,0,3) acc_year, substr(t.date, 0, 3) year,amount, balance, k_symbol
FROM ACCOUNT a, TRANS t, DISTRICT d
WHERE a.district_id = d.A1 and d.A2 = "Vyskov"
AND a.account_id = t.account_id


 * sqlite:///data_berka.db
Done.
Returning data to local variable dataset


In [ ]:
dataset.DataFrame().to_csv("dataset.csv")